runtime/compiler/control/JITServerHelpers.cpp

```diff
@@ -920,29 +920,41 @@ JITServerHelpers::postStreamFailure(OMRPortLibrary *portLibrary, TR::Compilation
    // If there was another networking problem (read/write error), the client stream
    // will be deleted, the socket will be closed and the client will be forced to
    // reconnect again. The client will discover if the server is down during the connection attempt.
-   if (connectionFailure && !retryConnectionImmediately)
+   if ((connectionFailure && !retryConnectionImmediately) || (!JITServer::CommunicationStream::shouldReadRetry()))
       {
-      if (TR::Options::getCmdLineOptions()->getVerboseOption(TR_VerboseJITServerConns))
+      if(connectionFailure)
          {
-         if (compInfo->getPersistentInfo()->getServerUID() != 0)
+         if (TR::Options::getCmdLineOptions()->getVerboseOption(TR_VerboseJITServerConns))
             {
-            TR_VerboseLog::writeLineLocked(TR_Vlog_JITServer,
+            if (compInfo->getPersistentInfo()->getServerUID() != 0)
+               {
+               TR_VerboseLog::writeLineLocked(TR_Vlog_JITServer,
                                            "t=%6u Lost connection to the server (serverUID=%llu). Next attempt in %llu ms.",
                                            (uint32_t)compInfo->getPersistentInfo()->getElapsedTime(),
                                            (unsigned long long)compInfo->getPersistentInfo()->getServerUID(),
                                            _waitTimeMs);
-            }
-         else // Was not connected to a server
-            {
-            TR_VerboseLog::writeLineLocked(TR_Vlog_JITServer,
+               }
+            else // Was not connected to a server
+               {
+               TR_VerboseLog::writeLineLocked(TR_Vlog_JITServer,
                                            "t=%6u Could not connect to a server. Next attempt in %llu ms.",
                                            (uint32_t)compInfo->getPersistentInfo()->getElapsedTime(),
                                            _waitTimeMs);
+               }
             }
+         compInfo->getPersistentInfo()->setServerUID(0);
+         _serverAvailable = false;
          }
-      compInfo->getPersistentInfo()->setServerUID(0);
-      _serverAvailable = false;
-
+      else
+         {
+         if (TR::Options::getCmdLineOptions()->getVerboseOption(TR_VerboseJITServerConns))
+            {
+              TR_VerboseLog::writeLineLocked(TR_Vlog_JITServer,
+                                           "t=%6u Multiple read failures. Next connection attempt in %llu ms.",
```

Comment:
Do we actually try to connect again? I see that `cannotPerformRemoteComp()` looks at `CommunicationStream::shouldReadRetry()` which looks at the number of failures and once it answers NO, it will always answer NO